central/auth/datastore/datastore_impl.go

```diff
@@ -27,16 +29,28 @@ func (d *datastoreImpl) AddAuthM2MConfig(ctx context.Context, config *storage.Au
 	if err := d.store.Upsert(ctx, config); err != nil {
 		return nil, err
 	}
+
+	if err := d.set.UpsertTokenExchanger(config); err != nil {
+		return nil, err
+	}
```

Comment:
>In this case, I would think of:

Create exchanger object. If that fails, return error.
Upsert a record in the database. If that fails, return error.
Add exchanger struct to the map. If it's just mutating Go map, that shouldn't fail (out of memory conditions can happen but if the process crashes at this point we know that exchanger is functional because we could create it in step (1)).

I think this combined with a shared mutex for the exchanger set and datastore seems like the best fit for the moment.

>You can query the config table by unique key (assuming you get the unique key as part of the input to the Exchange request) which should be very efficient.

This is the biggest friction which I _don't_ want to have. While theoretically possible to _require_ people to pass the ID of the config to use for the token exchange, this would feel like a _huge_ UX issue: suddenly you have to maintain the ID throughout your whole GitHub action workflows / potential future other places.

The goal of this is to reduce the maintenance overhead we currently have with API tokens (as well as improve the security aspect of the access), but requiring the ID in the request adds _some_ of that maintenance again.

Reply:
Updated the code to introduce a lock on the datastore.

Reply:
Next step: for @msugakov to verify. Resolving to not block the merge.